components/tidb_query/src/codec/convert.rs

```diff
@@ -284,159 +269,190 @@ impl ToInt for Real {
+        self.to_int(ctx, FieldTypeTp::LongLong)
     }
+}
 
-    fn to_uint(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<u64> {
-        let s = get_valid_utf8_prefix(ctx, self)?;
-        let vs = get_valid_int_prefix(ctx, s)?;
-        let val = vs.parse::<u64>();
-        match val {
-            Ok(val) => val.to_uint(ctx, tp),
-            Err(_) => {
-                ctx.handle_overflow_err(Error::overflow("BIGINT UNSIGNED", &vs))?;
-                let val = integer_unsigned_upper_bound(tp);
-                Ok(val)
-            }
-        }
+impl ConvertTo<u64> for f64 {
+    /// Port from TiDB's types.ConvertFloatToUint
+    fn convert(&self, ctx: &mut EvalContext) -> Result<u64> {
+        self.to_uint(ctx, FieldTypeTp::LongLong)
     }
 }
 
-impl ToInt for std::borrow::Cow<'_, [u8]> {
-    fn to_int(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<i64> {
-        self.as_ref().to_int(ctx, tp)
+impl ConvertTo<i64> for Real {
+    /// Port from TiDB's TiDB's types.ConvertFloatToInt
+    fn convert(&self, ctx: &mut EvalContext) -> Result<i64> {
+        self.into_inner().convert(ctx)
     }
+}
 
-    fn to_uint(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<u64> {
-        self.as_ref().to_uint(ctx, tp)
+impl ConvertTo<u64> for Real {
+    /// Port from TiDB's types.ConvertFloatToUint
+    fn convert(&self, ctx: &mut EvalContext) -> Result<u64> {
+        self.into_inner().convert(ctx)
     }
 }
 
-impl ToInt for Bytes {
-    fn to_int(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<i64> {
-        self.as_slice().to_int(ctx, tp)
+impl ConvertTo<i64> for &[u8] {
+    /// Port from TiDB's types.StrToInt
+    fn convert(&self, ctx: &mut EvalContext) -> Result<i64> {
+        let s = get_valid_utf8_prefix(ctx, self)?;
+        s.convert(ctx)
     }
+}
 
-    fn to_uint(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<u64> {
-        self.as_slice().to_uint(ctx, tp)
+impl ConvertTo<u64> for &[u8] {
+    /// Port from TiDB's types.StrToUint
+    fn convert(&self, ctx: &mut EvalContext) -> Result<u64> {
+        let s = get_valid_utf8_prefix(ctx, self)?;
+        s.convert(ctx)
     }
 }
 
-impl ToInt for Decimal {
-    #[inline]
-    fn to_int(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<i64> {
-        // TODO: avoid this clone
-        let dec = round_decimal_with_ctx(ctx, self.clone())?;
-        let val = dec.as_i64_with_ctx(ctx)?;
-        val.to_int(ctx, tp)
+impl ConvertTo<i64> for &str {
+    /// Port from TiDB's types.StrToInt
+    fn convert(&self, ctx: &mut EvalContext) -> Result<i64> {
+        let s = self.trim();
+        let s = get_valid_int_prefix(ctx, s)?;
+        match s.parse::<i64>() {
+            Ok(val) => Ok(val),
+            Err(_) => {
+                ctx.handle_overflow_err(Error::overflow("BIGINT", s.to_string()))?;
+                // in TiDB, the strconv.ParseInt return i64::MAX or i64::MIN when overflow,
```

Comment:
The comment seems redundant.

Reply:
No, I think this comment is necessary because we need to let the maintainer know why we return this value when overflow

Reply:
Prefer using:
```
// To make compatible with TiDB, `i64::MAX` or `i64::MIN` will be returned if overflow occurred, see [strconv.ParseInt](https://golang.org/pkg/strconv/#ParseInt)
```